<launch>
    <arg name="input_image" default="/camera/color/image_raw"/>
    <arg name="camera_info" default="/camera/color/camera_info"/>
    <arg name="model" default="vitl"/>
    <arg name="model_path" default="$(find depth_anything_ros)/trained_data/depth_anything_v2_metric_hypersim_$(arg model).engine"/>
    <arg name="depth_scale" default="0.5"/>
    <arg name="use_rainbow_colormap" default="true"/>
    
    <group ns="depth_anything">
        <!-- Nodelet manager -->
        <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager"
              output="screen" respawn="true">
            <param name="num_worker_threads" value="4"/>
        </node>

        <!-- Depth estimation node -->
        <node name="depth_estimation" pkg="depth_anything_ros" type="depth_estimation_node" output="screen">
            <remap from="~input_image" to="$(arg input_image)"/>
            <remap from="~depth_registered/image_rect" to="depth_registered/image_rect"/>
            
            <param name="model_path" value="$(arg model_path)"/>
            <param name="depth_scale" value="$(arg depth_scale)"/>
            <param name="use_rainbow_colormap" value="$(arg use_rainbow_colormap)"/>
        </node>

        <!-- Point cloud generation -->
        <node pkg="nodelet" type="nodelet" name="point_cloud_xyzrgb"
              args="load depth_image_proc/point_cloud_xyzrgb nodelet_manager" output="screen">
            <remap from="rgb/camera_info" to="$(arg camera_info)"/>
            <remap from="rgb/image_rect_color" to="$(arg input_image)"/>
            <remap from="depth_registered/image_rect" to="depth_registered/image_rect"/>
            <remap from="depth_registered/points" to="depth_registered/points"/>
            
            <param name="queue_size" value="10"/>
            <param name="approximate_sync" value="true"/>
        </node>
    </group>

    <!-- RViz for visualization -->
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find depth_anything_ros)/rviz/depth_anything.rviz"/>

</launch>
